arch-design/delfin/DistributedPerformanceMetricCollection.md

```diff
@@ -0,0 +1,314 @@
+ 
+- Fair scheduling:
+    
+    With the approach, jobs are failed scheduled by the Performance metric collection scheduler across the Task Manager process.
+ 
+Pros: 
+ 
+- Performance monitoring scheduler is stateless. Hence It would be easy to restart the scheduler in the leader switch scenario.
+ 
+Cons:
+ 
+- A book keeping of all scheduled jobs has to get maintained.
+- Maintain health check of each nodes  
+- The system needs to readjust locally scheduled jobs for each scale in or out.
+ 
+## Centrally managed/scheduled Jobs
+------------------------------------
+ 
+- The approach scheduler maintains a list of all requested performance monitoring jobs. 
+- The jobs are managed and scheduled in the leader Task Manager process. 
+- The jobs get triggered at a configured time interval from the leader Task Manager process and triggers a RPC call to Task manager in the configured time interval.
+- RPC call initiate a execution unit which request metrics from respective storages through Driver Manager, request metrics are further pushed to configured datastore using Exporter Interface
+ 
+![](./Resources/CentrallyScheduledJobs.jpg)
+ 
+ 
+### Analysis of centrally managed/scheduled Jobs
+- Fault tolerance: 
+    To achieve some node failure tolerance scenario, 
+    - Leader node failure require reelection of leader
+    - Need to ensure currently active metric collection retrigger for worker/leader node failure scenario
+             
+- Distributed metric collection:
+    
+    This is an inherent feature for the local scheduled job where incoming jobs are distributed among Task Manager processes and managed locally (inprocess).
+ 
+- Scalability:
+    As all worker nodes are stateless in nature, scaling is effortless.
+ 
+- Fair scheduling:
+    With the approach even though jobs are centrally managed but Metric collection tasks (heavy weight processing) are fairly scheduled across Task Manager processes.
+ 
+Pros:
+ 
+- It's easy to manage jobs as it is centrally managed.
+- A metric collection task (heavy weight task) is distributed. It's easy to scale out or scale in.
+- Task is always fairly scheduled.
+- Job cleanup is easy 
+ 
+Cons:
+ 
+- It might take more time to make Performance scheduling up in the loader switch scenario.
+ 
+ 
+## With the analysis, Its evident that
+- One leader required to manage incoming jobs for both distributed or centrally managed job scheduling. Hance leader election algorithm is required to manage leader nodes across distributed Task Manager processes.
+- Scaling is seamless in centrally managed job scheduler
+- Fault tolerance is easier to manage in centrally managed schedulers.
+- Task distribution is entact for both scheduling
+- Performance of scheduled job and metric collection is one key factor. 
+Need to evaluate both approaches.
+ 
+ 
+## Leader election
```

Comment:
Generic comment: As I understand from the points below, that (using any technology) if one of the leaders goes down, there will be a consensus created amongst other nodes to elect a new leader. From the description two things:
a) Will there be any heartbeat sort of mechanism to check the existence of a leader? (it may be implicit in the technologies, but if we bring it out over here it will help
b) Consider a scenario, the Leader node has network failure (not host/node down) and its communication to a few of the nodes is broken. Now the leader will keep updating and processing jobs while the other nodes may assume that leader is gone and they will go for election. How is this scenario handled?

Reply:
a) Yes, there is heartbeat mechanism for updating distributed lock. It is explained in [Leased based leader election](https://github.com/sodafoundation/architecture-analysis/blob/1746b0e79e6a86f72872d58baccb2b0a9b0bf870/arch-design/delfin/DistributedPerformanceMetricCollection.md#leased-based-leader-election).